### ISSUE TITLE
Training on my own dataset

### ISSUE THREAD

**Issue by skymaze** [charlesq34/pointnet#15]
Hello, I tired to train model on my own dataset witch generated by lidar, but the models have different numbers of points because of the distance. Are there any suggested algorithm to increase or decrease (normalize) the number of points?

**Comment by charlesq34** [comment]
Hi @skymaze 

You can set a MAX number of points during training, if there are less than MAX points, just replicate existing points should be fine -- since we use max that will not alter final result. If there are more than MAX, you can randomly subsample the points.

during test you can have a smaller batch size and even larger MAX to evaluate all points :)
or if there are too many you can compute result in two steps -- by firstly extracting point embedding and then aggregate.

Hope it helps!

Best,
Charles

**Comment by EndingCredits** [comment]
When batching presumably you need to mask the batch before pooling to ensure the extra missing points are not included?

I did something like this:
```
# Turns the input list into a single tensor, padding the outputs.
# Produces a mask of size batch_size x max_list_len x 1
def batchToArrays(input_list):
    max_len = 0
    out = []; seq_len = []; masks = []
    for i in input_list: max_len = max(len(i),max_len)
    for l in input_list:
        # Zero pad output
        out.append(np.pad(np.array(l,dtype=np.float32), ((0,max_len-len(l)),(0,0)), mode='constant'))
        # Create mask... (max_len x 1 vector with len(l) ones and rest zeros)
        # Extra dimension is left in for easy broadcasting
        masks.append(np.pad(np.array(np.ones((len(l),1)),dtype=np.float32), ((0,max_len-len(l)),(0,0)), mode='constant') )
return out, masks
```
and then:
```
def mask_and_pool(elems, mask, pool_type='max_pool'):
    # Use broadcasting to multiply
    masked_elems = tf.multiply(elems, mask)
    # Pool using max pooling
    embed = tf.reduce_max(masked_elems, axis=1)
return embed
```
My batch and num_points are set to None in the placeholder so I don't have to worry about the exact size, just about making sure everything has the same dimensions.

**Comment by charlesq34** [comment]
Hi @EndingCredits 

If that's max pooling then it won't matter in fact. Replicating points will not alter max pooled result.
If it's average pooling then you do need to add a mask.

**Comment by karenachiketc** [comment]
Hi,
Thank you for your work. The code is really helpful.
What is the `faceid` field I see in the .h5 file? I have attached an image for your reference. I'm using a program called vitables to view the .h5 file.

![screenshot from 2018-02-14 15-45-56](https://user-images.githubusercontent.com/13439811/36199131-a4c8c308-119e-11e8-9c0c-6c7dceab9e8b.png)

I understand that `label` value at location (0,0) is the label for `data` at the same location and the `normal` value corresponds to the value returned from LIDAR device while poing cloud generation(as I understand)  

Also if I pass my data to the helper function provided in the `data_prep_util.py` file, the generated .h5 file can be used for training right?

**Comment by kevindean87** [comment]
Hello @charlesq34.

I have been able to use the part segmentation algorithm you generated with my own LiDAR data and it works beautifully. After generating the model, I freeze the graph and optimize it. I later load it and try to use it. However, it is expecting a Tensor shape of (10, 2048, 3) or (batch_size, point_num, 3) for the pointclouds_ph. Obviously, I want a batch_size of 1, and I want to set point_num to however many points are in my data (variable from ~10,000 to ~30,000). When I run your test.py (by simply loading the checkpoint) works perfectly; but I need a frozen graph (for the way my company wants to implement Tensorflow). Is there a way to set up the training so that it is independent of specific batch_size and/or point_num? Thanks for any help!

Kevin Edward Dean
